hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java

```diff
@@ -6,139 +6,159 @@
  */
 package org.hibernate.internal.util.collections;
 
+import java.lang.reflect.Array;
 import java.util.ArrayDeque;
-import java.util.Deque;
-import java.util.Iterator;
+import java.util.Arrays;
+import java.util.NoSuchElementException;
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Function;
 
 /**
  * A general-purpose stack impl supporting null values.
+ * <p>
+ * Most of the implementation was inspired by {@link ArrayDeque} methods
+ * and follows the same logic.
  *
  * @param <T> The type of things stored in the stack
  *
  * @author Steve Ebersole
  * @author Sanne Grinovero
+ * @author Marco Belladelli
  */
 public final class StandardStack<T> implements Stack<T> {
+	private T[] elements;
+	private int head = 0;
+	private int tail = 0;
 
-	private ArrayDeque<T> internalStack;
-	private static final Object NULL_TOKEN = new Object();
-
-	public StandardStack() {
+	@SuppressWarnings("unchecked")
+	public StandardStack(Class<T> type) {
+		elements = (T[]) Array.newInstance( type, 8 );
 	}
 
-	public StandardStack(T initial) {
-		stackInstanceExpected().addFirst( initial );
+	public StandardStack(Class<T> type, T initial) {
+		this( type );
+		push( initial );
 	}
 
 	@Override
-	public void push(T newCurrent) {
-		T toStore = newCurrent;
-		if ( newCurrent == null ) {
-			toStore = (T) NULL_TOKEN;
-		}
-		stackInstanceExpected().addFirst( toStore );
-	}
-
-	private Deque<T> stackInstanceExpected() {
-		if ( internalStack == null ) {
-			//"7" picked to use 8, but skipping the odd initialCapacity method
-			internalStack = new ArrayDeque<>( 7 );
+	public void push(T e) {
+		head = dec( head );
+		elements[head] = e;
+		if ( head == tail ) {
+			grow();
```

Comment:
It's a bit surprising that we might end up growing the array right *after* having stored the last element.

Would be fantastic to avoid growing after a push, but rather make space if there is no space before a push?

I guess this also relates with the reason about why you needed to pre-allocate the array even on construction. Neither should be necessary.